Auth0 authenticator return 429

The service returns 500s for any request since the auth0 authenticator gets stuck in a "too many requests" state. This needs to be properly handled by the api

```
$ l0 service logs --tail 100 dimsio
dimsio
------
2017/08/01 20:07:44 [ERROR] Authenticator encountered an unexpected error: Invalid status code: 429
2017/08/01 20:07:50 [ERROR] Authenticator encountered an unexpected error: Invalid status code: 429
2017/08/01 20:07:55 [ERROR] Authenticator encountered an unexpected error: Invalid status code: 429
2017/08/01 20:08:14 [ERROR] Authenticator encountered an unexpected error: Invalid status code: 429
2017/08/01 20:08:18 [ERROR] Authenticator encountered an unexpected error: Invalid status code: 429
2017/08/01 20:08:20 [ERROR] Authenticator encountered an unexpected error: Invalid status code: 429
2017/08/01 20:08:24 [ERROR] Authenticator encountered an unexpected error: Invalid status code: 429
```

--- a/auth/auth0_manager.go
+++ b/auth/auth0_manager.go
@@ -2,6 +2,7 @@ package auth
 
 import (
 	"fmt"
+	"log"
 	"time"
 
 	"github.com/zpatrick/go-cache"
@@ -11,10 +12,14 @@ import (
 var timeMultiplier = 1
 
 type Auth0Manager struct {
-	clientID   string
-	connection string
-	client     *rclient.RestClient
-	cache      *cache.Cache
+	clientID           string
+	connection         string
+	client             *rclient.RestClient
+	cache              *cache.Cache
+	throttle           <-chan time.Time
+	throttleMin        int
+	throttleMultiplier int
+	lastThrottled      time.Time
 }
 
 type oauthReq struct {
@@ -36,16 +41,32 @@ const (
 	validAuthExpiry = 1 * time.Hour
 )
 
-func NewAuth0Manager(domain, clientID, connection string) *Auth0Manager {
+func NewAuth0Manager(domain, clientID, connection string, throttleMin int) *Auth0Manager {
 	return &Auth0Manager{
-		clientID:   clientID,
-		connection: connection,
-		client:     rclient.NewRestClient(domain),
-		cache:      cache.New(),
+		clientID:           clientID,
+		connection:         connection,
+		client:             rclient.NewRestClient(domain),
+		cache:              cache.New(),
+		throttle:           time.Tick(time.Millisecond * time.Duration(throttleMin)),
+		throttleMin:        throttleMin,
+		throttleMultiplier: throttleMin,
+		lastThrottled:      time.Now(),
 	}
 }
 
 func (a *Auth0Manager) Authenticate(username, password string) (bool, error) {
+	if a.throttleMultiplier > a.throttleMin && time.Since(a.lastThrottled) > 1*time.Minute {
+		a.throttleMultiplier = a.throttleMultiplier / 2
+		if a.throttleMultiplier < a.throttleMin {
+			a.throttleMultiplier = a.throttleMin
+		}
+
+		log.Printf("[INFO] 1 min since last throttle update, decreasing throttle to %v", time.Millisecond*time.Duration(a.throttleMultiplier))
+
+		a.throttle = time.Tick(time.Millisecond / time.Duration(a.throttleMultiplier))
+		a.lastThrottled = time.Now()
+	}
+
 	key := fmt.Sprintf("%s:%s", username, password)
 	var cachedStatus *authStatus
 	if result, exists := a.cache.Getf(key); exists {
@@ -57,6 +78,9 @@ func (a *Auth0Manager) Authenticate(username, password string) (bool, error) {
 	if cachedStatus.isValid {
 		return true, nil
 	}
+
+	// throttle to deal with rate limiting
+	<-a.throttle
 
 	// will only sleep if cachedStatus already exists with a penalty
 	time.Sleep(cachedStatus.penalty * time.Duration(timeMultiplier))
@@ -71,10 +95,23 @@ func (a *Auth0Manager) Authenticate(username, password string) (bool, error) {
 	}
 
 	if err := a.client.Post("/oauth/ro", req, nil); err != nil {
-		if err, ok := err.(*rclient.ResponseError); ok && err.Response.StatusCode == 401 {
+		if err, ok := err.(*rclient.ResponseError); ok && (err.Response.StatusCode == 401 || err.Response.StatusCode == 429) {
 			cachedStatus.penalty += time.Second
 			if cachedStatus.penalty > maxPenalty {
 				cachedStatus.penalty = maxPenalty
+			}
+
+			if err.Response.StatusCode == 429 {
+				if a.throttleMultiplier > 0 {
+					a.throttleMultiplier = a.throttleMultiplier * 2
+				} else {
+					a.throttleMultiplier = 1
+				}
+
+				log.Printf("[INFO] Too many requests, increasing throttle to %v", time.Millisecond*time.Duration(a.throttleMultiplier))
+
+				a.throttle = time.Tick(time.Millisecond * time.Duration(a.throttleMultiplier))
+				a.lastThrottled = time.Now()
 			}
 
 			a.cache.Add(key, cachedStatus)

--- a/auth/auth0_manager.go
+++ b/auth/auth0_manager.go
@@ -12,14 +12,11 @@ import (
 var timeMultiplier = 1
 
 type Auth0Manager struct {
-	clientID           string
-	connection         string
-	client             *rclient.RestClient
-	cache              *cache.Cache
-	throttle           <-chan time.Time
-	throttleMin        int
-	throttleMultiplier int
-	lastThrottled      time.Time
+	clientID   string
+	connection string
+	client     *rclient.RestClient
+	cache      *cache.Cache
+	throttle   <-chan time.Time
 }
 
 type oauthReq struct {
@@ -41,32 +38,17 @@ const (
 	validAuthExpiry = 1 * time.Hour
 )
 
-func NewAuth0Manager(domain, clientID, connection string, throttleMin int) *Auth0Manager {
+func NewAuth0Manager(domain, clientID, connection string, rateLimit time.Duration) *Auth0Manager {
 	return &Auth0Manager{
-		clientID:           clientID,
-		connection:         connection,
-		client:             rclient.NewRestClient(domain),
-		cache:              cache.New(),
-		throttle:           time.Tick(time.Millisecond * time.Duration(throttleMin)),
-		throttleMin:        throttleMin,
-		throttleMultiplier: throttleMin,
-		lastThrottled:      time.Now(),
+		clientID:   clientID,
+		connection: connection,
+		client:     rclient.NewRestClient(domain),
+		cache:      cache.New(),
+		throttle:   time.Tick(rateLimit),
 	}
 }
 
 func (a *Auth0Manager) Authenticate(username, password string) (bool, error) {
-	if a.throttleMultiplier > a.throttleMin && time.Since(a.lastThrottled) > 1*time.Minute {
-		a.throttleMultiplier = a.throttleMultiplier / 2
-		if a.throttleMultiplier < a.throttleMin {
-			a.throttleMultiplier = a.throttleMin
-		}
-
-		log.Printf("[INFO] 1 min since last throttle update, decreasing throttle to %v", time.Millisecond*time.Duration(a.throttleMultiplier))
-
-		a.throttle = time.Tick(time.Millisecond / time.Duration(a.throttleMultiplier))
-		a.lastThrottled = time.Now()
-	}
-
 	key := fmt.Sprintf("%s:%s", username, password)
 	var cachedStatus *authStatus
 	if result, exists := a.cache.Getf(key); exists {
@@ -79,12 +61,32 @@ func (a *Auth0Manager) Authenticate(username, password string) (bool, error) {
 		return true, nil
 	}
 
-	// throttle to deal with rate limiting
-	<-a.throttle
-
 	// will only sleep if cachedStatus already exists with a penalty
 	time.Sleep(cachedStatus.penalty * time.Duration(timeMultiplier))
 
+	isAuthenticated, err := a.authenticate(username, password)
+	if err != nil {
+		return false, err
+	}
+
+	if !isAuthenticated {
+		cachedStatus.penalty += time.Second
+		if cachedStatus.penalty > maxPenalty {
+			cachedStatus.penalty = maxPenalty
+		}
+
+		a.cache.Add(key, cachedStatus)
+		return false, nil
+	}
+
+	cachedStatus.isValid = true
+	cachedStatus.penalty = 0 * time.Second
+	a.cache.Addf(key, cachedStatus, validAuthExpiry)
+
+	return true, nil
+}
+
+func (a *Auth0Manager) authenticate(username, password string) (bool, error) {
 	req := oauthReq{
 		ClientID:   a.clientID,
 		Connection: a.connection,
@@ -94,35 +96,30 @@ func (a *Auth0Manager) Authenticate(username, password string) (bool, error) {
 		Scope:      "openid",
 	}
 
-	if err := a.client.Post("/oauth/ro", req, nil); err != nil {
-		if err, ok := err.(*rclient.ResponseError); ok && (err.Response.StatusCode == 401 || err.Response.StatusCode == 429) {
-			cachedStatus.penalty += time.Second
-			if cachedStatus.penalty > maxPenalty {
-				cachedStatus.penalty = maxPenalty
-			}
+	for backoff := time.Duration(0); true; backoff += time.Millisecond * 500 {
+		time.Sleep(backoff * time.Duration(timeMultiplier))
+		<-a.throttle
 
-			if err.Response.StatusCode == 429 {
-				if a.throttleMultiplier > 0 {
-					a.throttleMultiplier = a.throttleMultiplier * 2
-				} else {
-					a.throttleMultiplier = 1
+		if err := a.client.Post("/oauth/ro", req, nil); err != nil {
+			if err, ok := err.(*rclient.ResponseError); ok {
+				switch err.Response.StatusCode {
+				case 401:
+					log.Printf("[DEBUG] User '%s' sent invalid Auth0 credentials.", username)
+					return false, nil
+				case 429:
+					log.Printf("[DEBUG] Auth0 returned 429 response for user '%s'; retrying.", username)
+					continue
+				default:
+					return false, err
 				}
-
-				log.Printf("[INFO] Too many requests, increasing throttle to %v", time.Millisecond*time.Duration(a.throttleMultiplier))
-
-				a.throttle = time.Tick(time.Millisecond * time.Duration(a.throttleMultiplier))
-				a.lastThrottled = time.Now()
 			}
 
-			a.cache.Add(key, cachedStatus)
-			return false, nil
+			return false, err
 		}
 
-		return false, err
+		log.Printf("User '%s' sent valid Auth0 credentials.", username)
+		break
 	}
 
-	cachedStatus.isValid = true
-	cachedStatus.penalty = 0 * time.Second
-	a.cache.Addf(key, cachedStatus, validAuthExpiry)
 	return true, nil
 }

--- a/auth/auth0_manager.go
+++ b/auth/auth0_manager.go
@@ -96,7 +96,7 @@ func (a *Auth0Manager) authenticate(username, password string) (bool, error) {
 		Scope:      "openid",
 	}
 
-	for backoff := time.Duration(0); true; backoff += time.Millisecond * 500 {
+	for backoff := time.Duration(0); true; backoff += (time.Millisecond * 500) {
 		time.Sleep(backoff * time.Duration(timeMultiplier))
 		<-a.throttle
 
@@ -117,7 +117,7 @@ func (a *Auth0Manager) authenticate(username, password string) (bool, error) {
 			return false, err
 		}
 
-		log.Printf("User '%s' sent valid Auth0 credentials.", username)
+		log.Printf("[DEBUG] User '%s' sent valid Auth0 credentials.", username)
 		break
 	}
 

--- a/auth/auth0_manager_test.go
+++ b/auth/auth0_manager_test.go
@@ -110,3 +110,25 @@ func TestAuth0ManagerAuthenticate_NotValidCredsAreChecked(t *testing.T) {
 
 	assert.Equal(t, count, 2)
 }
+
+func TestAuth0ManagerAuthenticate_RetriesOn429(t *testing.T) {
+	count := 0
+
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		if count < 2 {
+			MarshalAndWrite(t, w, nil, 429)
+		} else {
+			MarshalAndWrite(t, w, nil, 200)
+		}
+
+		count++
+	}
+
+	auth0Manager, server := newAuth0ManagerAndServer(handler)
+	defer server.Close()
+
+	_, err := auth0Manager.Authenticate("", "")
+	if err != nil {
+		t.Fatal(err)
+	}
+}

--- a/auth/auth_test.go
+++ b/auth/auth_test.go
@@ -7,13 +7,14 @@ import (
 	"net/http"
 	"net/http/httptest"
 	"testing"
+	"time"
 )
 
 type Handler func(w http.ResponseWriter, r *http.Request)
 
 func newAuth0ManagerAndServer(handler Handler) (*Auth0Manager, *httptest.Server) {
 	server := httptest.NewServer(http.HandlerFunc(handler))
-	auth0Manager := NewAuth0Manager(server.URL, "", "", 500)
+	auth0Manager := NewAuth0Manager(server.URL, "", "", 500*time.Millisecond)
 
 	return auth0Manager, server
 }

--- a/auth/auth_test.go
+++ b/auth/auth_test.go
@@ -14,7 +14,7 @@ type Handler func(w http.ResponseWriter, r *http.Request)
 
 func newAuth0ManagerAndServer(handler Handler) (*Auth0Manager, *httptest.Server) {
 	server := httptest.NewServer(http.HandlerFunc(handler))
-	auth0Manager := NewAuth0Manager(server.URL, "", "", 500*time.Millisecond)
+	auth0Manager := NewAuth0Manager(server.URL, "", "", time.Second/2)
 
 	return auth0Manager, server
 }

--- a/auth/auth_test.go
+++ b/auth/auth_test.go
@@ -13,7 +13,7 @@ type Handler func(w http.ResponseWriter, r *http.Request)
 
 func newAuth0ManagerAndServer(handler Handler) (*Auth0Manager, *httptest.Server) {
 	server := httptest.NewServer(http.HandlerFunc(handler))
-	auth0Manager := NewAuth0Manager(server.URL, "", "")
+	auth0Manager := NewAuth0Manager(server.URL, "", "", 500)
 
 	return auth0Manager, server
 }

--- a/main.go
+++ b/main.go
@@ -99,7 +99,8 @@ func main() {
 		tokenManager := auth.NewDynamoTokenManager(c.String("dynamo-table"), dynamodb)
 		auth0Manager := auth.NewAuth0Manager(c.String("auth0-domain"),
 			c.String("auth0-client-id"),
-			c.String("auth0-connection"))
+			c.String("auth0-connection"),
+			500)
 
 		authenticator := auth.NewCompositeAuthenticator(tokenManager, auth0Manager)
 		proxy := proxy.NewECRProxy(c.String("registry-endpoint"))

--- a/main.go
+++ b/main.go
@@ -101,7 +101,7 @@ func main() {
 		auth0Manager := auth.NewAuth0Manager(c.String("auth0-domain"),
 			c.String("auth0-client-id"),
 			c.String("auth0-connection"),
-			time.Millisecond*500)
+			time.Second/2)
 
 		authenticator := auth.NewCompositeAuthenticator(tokenManager, auth0Manager)
 		proxy := proxy.NewECRProxy(c.String("registry-endpoint"))

--- a/main.go
+++ b/main.go
@@ -5,6 +5,7 @@ import (
 	"log"
 	"net/http"
 	"os"
+	"time"
 
 	"github.com/aws/aws-sdk-go/aws/credentials"
 	"github.com/aws/aws-sdk-go/aws/defaults"
@@ -100,7 +101,7 @@ func main() {
 		auth0Manager := auth.NewAuth0Manager(c.String("auth0-domain"),
 			c.String("auth0-client-id"),
 			c.String("auth0-connection"),
-			500)
+			time.Millisecond*500)
 
 		authenticator := auth.NewCompositeAuthenticator(tokenManager, auth0Manager)
 		proxy := proxy.NewECRProxy(c.String("registry-endpoint"))

